src/odemis/gui/cont/streams.py

```diff
@@ -2990,3 +2992,356 @@ def _updateScanStage(self, use):
+        # Delete all ROIs of the project
+        # .remove_roa_ctrl automatically removes itself from .roa_ctrls, so a for-loop doesn't work
+        while project_ctrl.roa_ctrls:
+            project_ctrl.remove_roa_ctrl(next(iter(project_ctrl.roa_ctrls.values())))
+
+        # Remove panel
+        self._project_bar.remove_project_panel(project_ctrl.panel)
+
+        # Remove controller from .project_ctrls list
+        self.project_ctrls = {key: val for key, val in self.project_ctrls.items() if val != project_ctrl}
+
+        # Remove model
+        self._tab_data_model.projects.value.remove(project_ctrl.model)
+
+        # Destroy ROAController object
+        del project_ctrl
+
+
+class FastEMProjectController(object):
+    """
+    Controller for a FastEM project. This class is responsible for the creation and maintenance
+    of ROIs belonging to the project.
+    During initialization, a panel is created and added to the project bar.
+    """
+
+    def __init__(self, name, colour, tab_data, project_bar, view_ctrl):
+        """
+        name (str): default name for the project
+        colour (str): hexadecimal colour code for the bounding box of the roas in the viewport
+        tab_data (FastEMAcquisitionGUIData): tab data model
+        project_bar (FastEMProjectBar): top-level panel containing all project panels
+        view_ctrl (FastEMAcquisitionViewport): viewport controller
+        """
+        self._tab_data = tab_data
+        self._project_bar = project_bar
+        self._view_ctrl = view_ctrl
+
+        self.roa_ctrls = {}  # dict int --> FastEMROAController
+        self.colour = colour
+        self.model = guimodel.FastEMProject(name)
+
+        # Create the panel and add it to the project bar. Subscribe to controls.
+        self.panel = FastEMProjectPanel(project_bar, name=name)
+        project_bar.add_project_panel(self.panel)
+        self.panel.btn_add_roi.Bind(wx.EVT_BUTTON, self._on_btn_roi)
+        self.panel.txt_ctrl.Bind(wx.EVT_TEXT, self._on_text)
+
+        # For ROA creation process
+        self._current_roa_ctrl = None
+
+    def _on_text(self, evt):
+        txt = evt.GetString()
+        logging.debug("Renaming project from %s to %s.", self.model.name.value, txt)
+        self.model.name.value = txt
+
+    def _on_btn_roi(self, _):
+        # Two-step process: Instantiate FastEM object here, but wait until first ROI is selected until
+        # further processing. The process can still be aborted by clicking in the viewport without dragging.
+        # In the callback to the ROI, the ROI creation will be completed or aborted.
+        self._project_bar.enable_buttons(False)
+
+        # Deactivate all ROAs
+        for roa_ctrl in self.roa_ctrls.values():
+            roa_ctrl.overlay.active.value = False
+
+        # Minimum index that has not yet been deleted
+        num = next(idx for idx, n in enumerate(sorted(self.roa_ctrls.keys()) + [0], 1) if idx != n)
```

Comment:
I think I see how it works, but it's quite hard to read. Maybe add "find the first index which is not in the existing indices".